core/server/master/src/main/java/alluxio/master/file/ListStatusPartial.java

```diff
@@ -0,0 +1,208 @@
+        // use the startAfter option, since this is the first listing
+        if (!partialOptions.get().getStartAfter().isEmpty()) {
+          String[] startAfter = PathUtils.getPathComponents(
+              partialOptions.get().getStartAfter());
+          ArrayList<String> startAfterList = new ArrayList<>(startAfter.length - 1);
+          startAfterList.addAll(Arrays.asList(startAfter).subList(1, startAfter.length));
+          return startAfterList;
+        }
+      }
+      // otherwise, start from the beginning of the listing
+      return Collections.emptyList();
+    }
+    // compute where to start from in each depth, we skip past the rootInodes, since that is
+    // where we start the traversal from
+    if (pathNames.size() <= rootPath.size()) {
+      return Collections.emptyList();
+    }
+    ArrayList<String> partialPath = new ArrayList<>(pathNames.size() - rootPath.size());
+    partialPath.addAll(pathNames.subList(rootPath.size(), pathNames.size()));
+    return partialPath;
+  }
+
+  /**
+   * If listing using a prefix and a partial path, this will compute whether the prefix
+   * exits in the partial path.
+   *
+   * @param partialPath the components of the path from where to start the listing
+   * @return the components of the prefix split by the delimiter /
+   * @throws InvalidPathException if the path in prefixComponents does not exist in partialPath
+   */
+  static List<String> checkPrefixListingPaths(
+      ListStatusContext context, List<String> partialPath)
+      throws InvalidPathException {
+    Optional<ListStatusPartialPOptions.Builder> pOptions = context.getPartialOptions();
+    if (!pOptions.isPresent()) {
+      return Collections.emptyList();
+    }
+    List<String> prefixComponents;
+    ListStatusPartialPOptions.Builder partialOptions = pOptions.get();
+    if (!partialOptions.getPrefix().isEmpty()) {
+      // compute the prefix as path components, removing the first empty string
+      prefixComponents = Arrays.stream(PathUtils.getPathComponents(
+              new AlluxioURI(
+                  partialOptions.getPrefix()).getPath())).skip(1)
+          .collect(Collectors.toList());
+    } else {
+      prefixComponents = Collections.emptyList();
+    }
+    // we only have to check the prefix if we are doing a partial listing,
+    // and we are not on the initial partial listing call
+    if (partialPath.isEmpty()
+        || !(partialOptions.hasOffset() && partialOptions.getOffset() != 0)) {
+      return prefixComponents;
+    }
+    // for each component the prefix must be the same as the partial path component
+    // except at the last component where the prefix must be contained in the partial path component
+    if (!PathUtils.hasPrefixComponentsCanBeLonger(partialPath, prefixComponents)) {
+      throw new InvalidPathException(
+          ExceptionMessage.PREFIX_DOES_NOT_MATCH_PATH
+              .getMessage(prefixComponents, partialPath));
+    }
```

Comment:
move the check up to fail fast?

Reply:
It is done here because the path and prefix components are generated and checked individually first.